### <div class="alert alert-info">Name: Dnyanai Surkutwar <br>Email: dsurkutwar@scu.edu  <br>Student ID: 00001586576 <br><center>Quiz#2<center></div>

In [1]:
## Import libaries 
import pandas as pd
import pickle 

In [2]:
## Import the csv file into dataset
d = pd.read_csv('songs.csv',encoding='latin-1') ## I had an issue reading the dataset in the right encoding, so decided to 
                                                ## convert it into utf-8 encoding. 
d.to_csv('songs.csv',index=False,encoding='utf-8') 
## Here, we know what our file encoding is and we are setting index and header as False as we do not want then in our dataFrame

## Reading the csv file into a dataframe
dataset = pd.read_csv('songs.csv',encoding='utf-8')
dataset.head()

year                           songtitle         artistname  \
0  2010  This Is the House That Doubt Built  A Day to Remember   
1  2010                     Sticks & Bricks  A Day to Remember   
2  2010                          All I Want  A Day to Remember   
3  2010                    It's Complicated  A Day to Remember   
4  2010                           2nd Sucks  A Day to Remember   

               songID            artistID  timesignature  \
0  SOBGGAB12C5664F054  AROBSHL1187B9AFB01              3   
1  SOPAQHU1315CD47F31  AROBSHL1187B9AFB01              4   
2  SOOIZOU1376E7C6386  AROBSHL1187B9AFB01              4   
3  SODRYWD1315CD49DBE  AROBSHL1187B9AFB01              4   
4  SOICMQB1315CD46EE3  AROBSHL1187B9AFB01              4   

   timesignature_confidence  loudness    tempo  tempo_confidence  ...  \
0                     0.853    -4.262   91.525             0.953  ...   
1                     1.000    -4.051  140.048             0.921  ...   
2                     1.000    -3.571  160.512             0.489  ...   
3                     1.000    -3.815   97.525             0.794  ...   
4                     0.788    -4.707  140.053             0.286  ...   

   timbre_7_max  timbre_8_min  timbre_8_max  timbre_9_min  timbre_9_max  \
0        82.475       -52.025        39.116       -35.368        71.642   
1       106.918       -61.320        35.378       -81.928        74.574   
2        80.621       -59.773        45.979       -46.293        59.904   
3        96.675       -78.660        41.088       -49.194        95.440   
4       110.332       -56.450        37.555       -48.588        67.570   

   timbre_10_min  timbre_10_max  timbre_11_min  timbre_11_max  Top10  
0       -126.440         18.658        -44.770         25.989      0  
1       -103.808        121.935        -38.892         22.513      0  
2       -108.313         33.300        -43.733         25.744      0  
3       -102.676         46.422        -59.439         37.082      0  
4        -52.796         22.888        -50.414         32.758      0  

[5 rows x 39 columns]

### Performing EDA on the dataset: 

In [3]:
dataset.describe()
## The total number of songs are 7574.
## 50% of the songs are from year 2002.
## 75% of the songs are from year 2006.

year  timesignature  timesignature_confidence     loudness  \
count  7574.000000    7574.000000               7574.000000  7574.000000   
mean   2001.455902       3.893979                  0.853307    -8.817262   
std       5.815223       0.532561                  0.243748     4.377079   
min    1990.000000       0.000000                  0.000000   -42.451000   
25%    1997.000000       4.000000                  0.819250   -10.847000   
50%    2002.000000       4.000000                  0.979000    -7.649000   
75%    2006.000000       4.000000                  1.000000    -5.640000   
max    2010.000000       7.000000                  1.000000     1.305000   

             tempo  tempo_confidence          key  key_confidence  \
count  7574.000000       7574.000000  7574.000000     7574.000000   
mean    107.348267          0.622866     5.384605        0.433843   
std      24.665030          0.304656     3.572579        0.273826   
min       0.000000          0.000000     0.000000        0.000000   
25%      88.860250          0.372000     2.000000        0.204000   
50%     103.268000          0.701500     6.000000        0.451500   
75%     124.800500          0.892000     9.000000        0.646000   
max     244.307000          1.000000    11.000000        1.000000   

            energy        pitch  ...  timbre_7_max  timbre_8_min  \
count  7574.000000  7574.000000  ...   7574.000000   7574.000000   
mean      0.675471     0.010817  ...     95.653221    -63.703984   
std       0.243473     0.013671  ...     26.859726     15.970244   
min       0.000020     0.000000  ...     15.699000   -158.756000   
25%       0.500138     0.003000  ...     76.498000    -73.050750   
50%       0.718160     0.007000  ...     94.632500    -62.661500   
75%       0.887396     0.014000  ...    112.708250    -52.982750   
max       0.998492     0.541000  ...    214.819000     -2.382000   

       timbre_8_max  timbre_9_min  timbre_9_max  timbre_10_min  timbre_10_max  \
count   7574.000000   7574.000000   7574.000000    7574.000000    7574.000000   
mean      50.057489    -59.515016     68.028067     -87.340055      55.520722   
std       14.463524     17.190527     20.381321      30.464274      23.549473   
min      -25.955000   -149.507000      8.415000    -208.819000      -6.359000   
25%       40.584500    -70.282500     53.037250    -105.130250      39.196000   
50%       49.220000    -58.650000     65.935000     -83.074000      50.895000   
75%       58.462000    -47.699750     81.267000     -64.518000      66.593000   
max      144.985000      1.140000    161.518000     -10.640000     192.417000   

       timbre_11_min  timbre_11_max        Top10  
count    7574.000000    7574.000000  7574.000000  
mean      -50.868457      47.490774     0.147742  
std        11.701530      12.463966     0.354868  
min      -145.599000       7.200000     0.000000  
25%       -58.058000      38.975250     0.000000  
50%       -50.892500      46.437500     0.000000  
75%       -43.292250      55.030250     0.000000  
max        -6.497000     110.272000     1.000000  

[8 rows x 35 columns]

In [4]:
## Lets see the dataset info which shows use the non-null values in the dataframe columns 
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7574 entries, 0 to 7573
Data columns (total 39 columns):
year                        7574 non-null int64
songtitle                   7574 non-null object
artistname                  7574 non-null object
songID                      7574 non-null object
artistID                    7574 non-null object
timesignature               7574 non-null int64
timesignature_confidence    7574 non-null float64
loudness                    7574 non-null float64
tempo                       7574 non-null float64
tempo_confidence            7574 non-null float64
key                         7574 non-null int64
key_confidence              7574 non-null float64
energy                      7574 non-null float64
pitch                       7574 non-null float64
timbre_0_min                7574 non-null float64
timbre_0_max                7574 non-null float64
timbre_1_min                7574 non-null float64
timbre_1_max                7574 non-null float64
tim

In [5]:
## Checking if there are null values in the dataframe columns
dataset.isnull().sum()

year                        0
songtitle                   0
artistname                  0
songID                      0
artistID                    0
timesignature               0
timesignature_confidence    0
loudness                    0
tempo                       0
tempo_confidence            0
key                         0
key_confidence              0
energy                      0
pitch                       0
timbre_0_min                0
timbre_0_max                0
timbre_1_min                0
timbre_1_max                0
timbre_2_min                0
timbre_2_max                0
timbre_3_min                0
timbre_3_max                0
timbre_4_min                0
timbre_4_max                0
timbre_5_min                0
timbre_5_max                0
timbre_6_min                0
timbre_6_max                0
timbre_7_min                0
timbre_7_max                0
timbre_8_min                0
timbre_8_max                0
timbre_9_min                0
timbre_9_m

### Q1.1) 

In [6]:
print('The total number of songs in total are :',dataset.songtitle.count()) ## count gives the count of non-null values in 
                                                                            ## the dataframe

The total number of songs in total are : 7574


In [7]:
dataset.Top10.value_counts() ## value_counts() gives us the count of unique values in the dataframe

0    6455
1    1119
Name: Top10, dtype: int64

In [8]:
dataset.Top10[dataset.Top10==1].count() ##Finding out which songs belong to Top10 column by masking the data 

1119

In [9]:
print('The number of Top10 songs are: ',dataset.Top10[dataset.Top10==1].count()) 

The number of Top10 songs are:  1119


In [10]:
print('The number of non-Top10 songs are: ',dataset.Top10[~(dataset.Top10==1)].count()) ## Finding out the songs who do not
                                                                                        ## belong to Top10

The number of non-Top10 songs are:  6455


### <div class = 'alert alert-info'>Answer 1.1: <br><ol><li> The total number of songs in total are : 7574</li><br><li>The number of Top10 songs are:  1119</li><br><li>The number of non-Top10 songs are:  6455</li></ol></div>

In [11]:
## Pickling the file so that I can use it in another file for the next questions, pickling will help me start off directly
## on the currrent dataset instead of initializing it again
dataset.to_pickle('dataset.pickle')

### Q1.2) 

In [12]:
dataset.columns

Index(['year', 'songtitle', 'artistname', 'songID', 'artistID',
       'timesignature', 'timesignature_confidence', 'loudness', 'tempo',
       'tempo_confidence', 'key', 'key_confidence', 'energy', 'pitch',
       'timbre_0_min', 'timbre_0_max', 'timbre_1_min', 'timbre_1_max',
       'timbre_2_min', 'timbre_2_max', 'timbre_3_min', 'timbre_3_max',
       'timbre_4_min', 'timbre_4_max', 'timbre_5_min', 'timbre_5_max',
       'timbre_6_min', 'timbre_6_max', 'timbre_7_min', 'timbre_7_max',
       'timbre_8_min', 'timbre_8_max', 'timbre_9_min', 'timbre_9_max',
       'timbre_10_min', 'timbre_10_max', 'timbre_11_min', 'timbre_11_max',
       'Top10'],
      dtype='object')

In [20]:
dataset.artistname[dataset.artistname=='Michael Jackson'].count()

18

### <div class = 'alert alert-info'>Answer 1.2:<br>The artist named "Michael Jackson" owns 18 songs.</div>

In [21]:
dataset.columns

Index(['year', 'songtitle', 'artistname', 'songID', 'artistID',
       'timesignature', 'timesignature_confidence', 'loudness', 'tempo',
       'tempo_confidence', 'key', 'key_confidence', 'energy', 'pitch',
       'timbre_0_min', 'timbre_0_max', 'timbre_1_min', 'timbre_1_max',
       'timbre_2_min', 'timbre_2_max', 'timbre_3_min', 'timbre_3_max',
       'timbre_4_min', 'timbre_4_max', 'timbre_5_min', 'timbre_5_max',
       'timbre_6_min', 'timbre_6_max', 'timbre_7_min', 'timbre_7_max',
       'timbre_8_min', 'timbre_8_max', 'timbre_9_min', 'timbre_9_max',
       'timbre_10_min', 'timbre_10_max', 'timbre_11_min', 'timbre_11_max',
       'Top10'],
      dtype='object')